src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,141 @@
+        {
+            Manifest.Features = ContractFeatures.HasStorage | ContractFeatures.Payable;
+        }
+
+        [ContractMethod(0, CallFlags.AllowModifyStates)]
+        private void Finish(ApplicationEngine engine)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            OracleResponse response = tx.Attributes.OfType<OracleResponse>().First();
+            StorageKey key = CreateStorageKey(Prefix_Request).Add(response.Id);
+            OracleRequest request = engine.Snapshot.Storages[key].GetInteroperable<OracleRequest>();
+            StackItem userData = BinarySerializer.Deserialize(request.UserData, engine.MaxStackSize, engine.MaxItemSize, engine.ReferenceCounter);
+            engine.CallFromNativeContract(null, request.CallbackContract, request.CallbackMethod, request.Url, userData, response.Success, response.Result);
+        }
+
+        [ContractMethod(0_01000000, CallFlags.AllowStates)]
+        public ECPoint[] GetOracleNodes(StoreView snapshot)
+        {
+            return snapshot.Storages[CreateStorageKey(Prefix_NodeList)].GetInteroperable<NodeList>().ToArray();
+        }
+
+        public OracleRequest GetRequest(StoreView snapshot, ulong id)
+        {
+            return snapshot.Storages.TryGet(CreateStorageKey(Prefix_Request).Add(id))?.GetInteroperable<OracleRequest>();
+        }
+
+        public IEnumerable<OracleRequest> GetRequests(StoreView snapshot)
+        {
+            return snapshot.Storages.Find(new byte[] { Prefix_Request }).Select(p => p.Value.GetInteroperable<OracleRequest>());
+        }
+
+        public IEnumerable<OracleRequest> GetRequestsByUrl(StoreView snapshot, string url)
+        {
+            IdList list = snapshot.Storages.TryGet(CreateStorageKey(Prefix_IdList).Add(GetUrlHash(url)))?.GetInteroperable<IdList>();
+            if (list is null) yield break;
+            foreach (ulong id in list)
+                yield return snapshot.Storages[CreateStorageKey(Prefix_Request).Add(id)].GetInteroperable<OracleRequest>();
+        }
+
+        private static byte[] GetUrlHash(string url)
+        {
+            return Crypto.Hash160(Utility.StrictUTF8.GetBytes(url));
+        }
+
+        internal override void Initialize(ApplicationEngine engine)
+        {
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_NodeList), new StorageItem(new NodeList()));
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RequestId), new StorageItem(BitConverter.GetBytes(0ul)));
+        }
+
+        private void MintGasForOracleNode(ApplicationEngine engine, ulong id)
+        {
+            ECPoint[] nodes = GetOracleNodes(engine.Snapshot);
+            int i = (int)(id % (ulong)nodes.Length);
+            UInt160 account = Contract.CreateSignatureRedeemScript(nodes[i]).ToScriptHash();
+            GAS.Mint(engine, account, OracleRequestPrice);
+        }
+
+        protected override void PostPersist(ApplicationEngine engine)
```

Comment:
It's not better to move the TX to the last position?

Reply:
What do you mean?

Reply:
If we move the responseTx to the last transaction of the block, we can avoid `postPersist`

Reply:
We can't. `PostPersist()` will be executed even if the response callback fails.